protocols.py

```diff
@@ -1,50 +1,104 @@
-                         HighLevelILVar, Constant, Function, HighLevelILVarSsa, HighLevelILVarInitSsa,
-                         TypeFieldReference, bundled_plugin_path, log_info, log_warn, log_alert)
-from typing import Optional, Tuple
+                         HighLevelILVar, Constant, Function, HighLevelILVarSsa, HighLevelILVarInitSsa, show_message_box,
+                         TypeFieldReference, bundled_plugin_path, log_info, log_warn, log_alert, user_plugin_path)
+from binaryninja.enums import MessageBoxButtonSet
+from typing import Optional, Tuple, List
 import os
 import sys
 import struct
 
 protocols = None
 
-def init_protocol_mapping():
-    # Parse EFI definitions only once
-    global protocols
-    if protocols is not None:
-        return True
 
-    # Find the EFI type definition file within the Binary Ninja installation
+def init_protocol_mapping(bv: BinaryView) -> bool:
+    """
+    Init protocol mappings, this function will parse bundled efi types and user provided types (if applicable)
+    If the user-provided mapping files contain errors, it will show a message box containing error position.
+
+    :param bv: Binary view
+    :return: bool
+    """
+    global protocols
     if sys.platform == "darwin":
-        efi_def_path = os.path.join(bundled_plugin_path(), "..", "..", "Resources", "types", "efi.c")
+        bundled_efi_path = os.path.join(bundled_plugin_path(), "..", "..", "Resources", "types", "efi.c")
+        user_efi_path = os.path.join(user_plugin_path(), "..", "types", "efi.c")
     else:
-        efi_def_path = os.path.join(bundled_plugin_path(), "..", "types", "efi.c")
+        bundled_efi_path = os.path.join(bundled_plugin_path(), "..", "types", "efi.c")
+        user_efi_path = os.path.join(user_plugin_path(), "..", "types", "efi.c")
 
-    # Try to read the EFI type definitions. This may not exist on older versions of Binary Ninja.
     try:
-        efi_defs = open(efi_def_path, "r").readlines()
-    except:
-        log_alert(f"Could not open EFI type definition file at '{efi_def_path}'. Your version of Binary Ninja may be out of date. Please update to version 3.5.4331 or higher.")
+        with open(bundled_efi_path, "r") as f:
+            bundled_efi_defs = f.readlines()
+    except FileNotFoundError:
+        log_alert(f"Could not open EFI type definition file at '{bundled_efi_path}'. Your version of Binary Ninja may be out of date. Please update to version 3.5.4331 or higher.")
         return False
 
+    bundled_protocols = parse_protocol_mapping(bundled_efi_defs)
+    protocols = bundled_protocols
+
+    if os.path.exists(user_efi_path):
+        user_platform_path = os.path.abspath(os.path.join(user_plugin_path(), "..", "types", "platform", f"{bv.platform.name}.c"))
+        continue_text = "\nContinue without user-provided protocol bindings?"
+        if not os.path.exists(user_platform_path):
```

Comment:
Most of our users are just going to want to put all of their types in their `efi.c` file and are unlikely to have platform-specific types. I don't think we need this dialog and we should continue as long as they have an `efi.c` file. We should also check if they have a `types/platform/<platform type>.c` file and parse it if it exists. But I don't think we should prompt the user.

Also please document how to supply custom EFI types with GUIDs to the README so our users know how to do it.